FEAT: cross-posting to Mastodon

alifeee opened the pull request:
add `Mastodon` settings page and other logic in prep for cross-posting to mastodon

Done:

- modify user database table with parameters
- create settings page
- create test button on settings page
- add "post to mastodon" checkbox to log modal
- pass "post to mastodon" bool to movie play/watchlist logging PHP

to-do:

- create job to post to mastodon
- trigger job
- test

## context

Mastodon allows API access to accounts on https://mastodon.social/settings/applications/

<img width="1179" height="817" alt="image" src="https://github.com/user-attachments/assets/2d65d2a6-a4f9-4bd3-85cf-b534431cebf9" />

we can post statuses to Mastodon with:

```bash
TOKEN="8uqah2tr78hwauhfguihweauihguihesuigh973eh7u8ighe"
curl "https://mastodon.social/api/v1/statuses" \
  -H "Authorization: Bearer ${TOKEN}" \
  -H "Idempotency-Key: ${RANDOM}" \
  -F "status=test post using the API: ${RANDOM}" \
  -F "visibility=private"
```

## this PR

this PR adds this settings page `/settings/integrations/mastodon`, where you can enable cross-posting, input an API access token, input a username (which is used for the domain), choose what post type should be used, and enable automatic posting (see below)

<details><summary>the names of these settings are inconsistent, expand:</summary>

(due to so many files needing to be edited and the structure/nature of the application)

```text
in SQL
mastodon_xpost_enabled
mastodon_xpost_automatic
mastodon_xpost_username
mastodon_xpost_access_token
mastodon_xpost_post_visibility

in PHP
$mastodonXPostEnabled,
$mastodonXPostAutomatic,
$mastodonXPostUsername,
$mastodonXPostAccessToken,
$mastodonXPostPostVisibility,

in Twig        
mastodonEnable
mastodonOnByDefault
mastodonUsername
mastodonVisibility
mastodonAccessToken
```

</details>

<img width="734" height="877" alt="image" src="https://github.com/user-attachments/assets/04b5d451-405d-4209-917c-e24b12e0d0c0" />

then, on the log play modal, we get a checkbox (only visible if cross posting enabled, default value is from automatic)

<img width="513" height="385" alt="image" src="https://github.com/user-attachments/assets/0baeacd5-d072-4840-94ae-b13aed385226" />

## notes

### on jobs

currently, the settings are implemented, but the logic is not.

`WatchlistController->addMovieToWatchlist` and `HistoryController->logMovie` should create a job to post a Mastodon status. I've not done this yet. Jobs are confusing.

it also seems to me that: `movary-job-processing` runs every 15 seconds and processes the oldest job. this is no good if the rate of job creation ever goes above 1 every 15 seconds

### on watchlist adding

I have no plans to currently trigger Mastodon from the movie page as "add to watchlist" button uses completely different logic to the one on the log play modal (sad)

### on jobs (again)

is creating a job the best way of doing this ? I'm not sure… maybe @leepeuker would know more.

leepeuker commented:
@alifeee please have a look at my changes and check if everything works as expected

there are mainly two open TODOs: generating the actual Mastodon post messages (see in the `MastodonPostWatchlistService` and `MastodonPostPlayService` classes). feel free to continue from here, the rest looks fine to me and should work now

alifeee commented on src/Service/Mastodon/MastodonApi.php:
```diff
@@ -0,0 +1,49 @@
+            throw new RuntimeException('Could not extract domain from username: ' . $username);
+        }
+
+        $requestUrl = 'https://' . $username . '/api/v1/statuses';
```
is this supposed to be `$domain` ?

alifeee commented on src/Service/Mastodon/MastodonApi.php:
```diff
@@ -0,0 +1,49 @@
+            'connect_timeout' => self::DEFAULT_TIMEOUT,
+            'headers' => [
+                'Authorization' => 'Bearer ' . $apiToken,
+                'Idempotency-Key' => Uuid::uuid4()->toString(),
```
perhaps this idempotency key should be a hash of `$message`, but I do not think it matters so much

leepeuker replied:
I checked the Mastodon documentation, they mentioned UUIDs as examples for possible values, that is why I implement it this way.

leepeuker commented:
> > HTTP referer not being allowed might have screwed my local dev environment but I will figure it out and look at finishing this off
> 
> Can you give me more details about this? What exactly is your issue? I am not aware of open issues with the HTTP referer?

nvm I could reproduce the issue! I have fixed it by not using the HTTP referer and hardcoding the redirect url, which is probably better any way as we always want to redirect to the settings page afterwards

alifeee commented:
I changed the messages, mainly I just added an URL to the movie page

They look like this (with https://github.com/leepeuker/movary/pull/722 the preview will show the movie poster and a better description of Movary)

<img width="614" height="354" alt="image" src="https://github.com/user-attachments/assets/14532961-052b-4317-b98a-ea1b38a23630" />

<img width="614" height="354" alt="image" src="https://github.com/user-attachments/assets/b7479294-88de-47ea-9973-af30e6589ff1" />

leepeuker commented:
Note: Update of the docs with the new integration is missing

alifeee commented:
relevant PRs in no particular order

https://github.com/leepeuker/movary/pull/732 https://github.com/leepeuker/movary/pull/729 https://github.com/leepeuker/movary/pull/726 https://github.com/leepeuker/movary/pull/725 https://github.com/leepeuker/movary/pull/722 https://github.com/leepeuker/movary/pull/728 https://github.com/leepeuker/movary/pull/735
